platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/type/ModuleRejectedFailure.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.resolution.failure.type;
+
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.ModuleResolveState;
+
+public final class ModuleRejectedFailure extends AbstractComponentSelectionFailure {
+    private final ModuleResolveState moduleResolveState;
+
+    public ModuleRejectedFailure(ModuleIdentifier moduleIdentifier, ModuleResolveState moduleResolveState) {
+        super(moduleIdentifier);
+        this.moduleResolveState = moduleResolveState;
```

Comment:
We should not retain instances of `ModuleResolveState` in failures. This is an internal dependency resolution engine type